storage/innobase/row/row0mysql.cc

```diff
@@ -2213,9 +2297,17 @@ row_update_cascade_for_mysql(
 
 	const trx_t* trx = thr_get_trx(thr);
 
-	bool vers_set_fields = node->table->versioned()
-				  && (node->is_delete == PLAIN_DELETE
-				      || node->update->affects_versioned());
+	bool vers_set_fields = table->versioned()
+				&& (node->is_delete == PLAIN_DELETE
+				    || node->update->affects_versioned());
+
+	if (table->versioned() && !node->is_delete
+	    && node->update->affects_versioned()) {
+		dberr_t err = row_update_versioned_insert(thr, node);
```

Comment:
Position in `node->pcur` is changed after `row_update_versioned_insert()`. Why it's not important?

Reply:
Because `node` is not running at that moment.

`node->pcur` got changed because `row_update_versioned_insert()` restores it. That is not harmful.

`node` restores its `pcur` later when it needs it:
```c++
#0  btr_pcur_restore_position_func (latch_mode=2, cursor=0x7fffa0104850, file=0x165d0a0 "/home/kevg/work/mariadb/storage/innobase/row/row0upd.cc", line=3102, mtr=0x7fffedc99a78) at btr0pcur.cc:235
#1  row_upd_clust_step (node=0x7fffa015b858, thr=0x7fffa015ae10) at row0upd.cc:3102
#2  row_upd (node=0x7fffa015b858, thr=0x7fffa015ae10) at row0upd.cc:3288
#3  row_upd_step (thr=0x7fffa015ae10) at row0upd.cc:3432
#4  row_update_cascade_for_mysql (thr=0x7fffa015ae10, node=0x7fffa015b858, table=0x7fffa0038d08) at row0mysql.cc:2324
#5  row_ins_foreign_check_on_constraint (thr=0x7fffa015ae10, foreign=0x7fffa003b2a8, pcur=0x7fffedc9b198, entry=0x7fffa018e960, mtr=0x7fffedc9ac80) at row0ins.cc:1419
#6  row_ins_check_foreign_constraint (check_ref=0, foreign=0x7fffa003b2a8, table=0x7fffa0156d38, entry=0x7fffa018e960, thr=0x7fffa015ae10) at row0ins.cc:1796
#7  row_upd_check_references_constraints (node=0x7fffa015ab18, pcur=0x7fffa0032d50, table=0x7fffa0156d38, index=0x7fffa01848f8, offsets=0x7fffa01673a0, thr=0x7fffa015ae10, mtr=0x7fffedc9bda8) at row0upd.cc:326
#8  row_upd_clust_rec_by_insert (node=0x7fffa015ab18, index=0x7fffa01848f8, thr=0x7fffa015ae10, referenced=1, foreign=false, mtr=0x7fffedc9bda8) at row0upd.cc:2758
#9  row_upd_clust_step (node=0x7fffa015ab18, thr=0x7fffa015ae10) at row0upd.cc:3212
#10 row_upd (node=0x7fffa015ab18, thr=0x7fffa015ae10) at row0upd.cc:3288
#11 row_upd_step (thr=0x7fffa015ae10) at row0upd.cc:3432
#12 row_update_for_mysql (prebuilt=0x7fffa015a018) at row0mysql.cc:1940
#13 ha_innobase::update_row (this=0x7fffa0037230, old_row=0x7fffa01573e8 "\377\001", new_row=0x7fffa01573e0 "\377\v") at ha_innodb.cc:9044
#14 handler::ha_update_row (this=0x7fffa0037230, old_data=0x7fffa01573e8 "\377\001", new_data=0x7fffa01573e0 "\377\v") at handler.cc:6245
#15 mysql_update (thd=0x7fffa0000ce8, table_list=0x7fffa00140b8, fields=List<Item> with 2 elements, values=List<Item> with 2 elements, conds=0x0, order_num=0, order=0x0, limit=18446744073709551615, handle_duplicates=DUP_ERROR, ignore=false, found_return=0x7fffedc9ed28, updated_return=0x7fffedc9ed20) at sql_update.cc:942
#16 mysql_execute_command (thd=0x7fffa0000ce8) at sql_parse.cc:4565
#17 mysql_parse (thd=0x7fffa0000ce8, rawbuf=0x7fffa0013fb0 "update parent set id = 11, value = value + 1", length=44, parser_state=0x7fffedca0400, is_com_multi=false, is_next_command=false) at sql_parse.cc:8009
```

And it even can restore it multiple times in a loop inside `row_update_cascade_for_mysql()`:
```c++
	for (;;) {
		if (vers_set_fields) {
			node->vers_set_fields(trx);
		}

		thr->run_node = node;
		thr->prev_node = node;

		DEBUG_SYNC_C("foreign_constraint_update_cascade");
		{
			TABLE *mysql_table = thr->prebuilt->m_mysql_table;
			thr->prebuilt->m_mysql_table = NULL;
			row_upd_step(thr);
			thr->prebuilt->m_mysql_table = mysql_table;
		}
```